# Leitura do Dataset e limpeza


In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import plotly.express as px

LINK PARA O DATASET https://www.kaggle.com/datasets/psychewolf/dota-2-heroes-dataset?select=heroes_data.csv

importada as bibliotecas necessárias

<h3>Leitura do csv e visualização inicial</h3>


In [2]:
df = pd.read_csv('./data/heroes_data.csv',sep=',')
df.sample(2)

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



Unnamed: 0   id                      name localized_name primary_attr  \
29           29   31        npc_dota_hero_lich           Lich          int   
115         115  121  npc_dota_hero_grimstroke     Grimstroke          int   

    attack_type                                       roles  \
29       Ranged                        ['Support', 'Nuker']   
115      Ranged  ['Support', 'Nuker', 'Disabler', 'Escape']   

                                                img  \
29         /apps/dota2/images/heroes/lich_full.png?   
115  /apps/dota2/images/heroes/grimstroke_full.png?   

                                              icon  base_health  ...  5_win  \
29         /apps/dota2/images/heroes/lich_icon.png          200  ...  13135   
115  /apps/dota2/images/heroes/grimstroke_icon.png          200  ...  13142   

     6_pick  6_win  7_pick  7_win  8_pick  8_win    winrate  lane  \
29    13112   6518    6321   3067    1713    827  48.469007     5   
115   13809   6571    6162   2900    2168    981  46.590636     5   

        main_role  
29   hard_support  
115  hard_support  

[2 rows x 55 columns]

In [3]:
df.columns

Index(['Unnamed: 0', 'id', 'name', 'localized_name', 'primary_attr',
       'attack_type', 'roles', 'img', 'icon', 'base_health',
       'base_health_regen', 'base_mana', 'base_mana_regen', 'base_armor',
       'base_mr', 'base_attack_min', 'base_attack_max', 'base_str', 'base_agi',
       'base_int', 'str_gain', 'agi_gain', 'int_gain', 'attack_range',
       'projectile_speed', 'attack_rate', 'move_speed', 'turn_rate',
       'cm_enabled', 'legs', 'hero_id', 'turbo_picks', 'turbo_wins', 'pro_ban',
       'pro_win', 'pro_pick', '1_pick', '1_win', '2_pick', '2_win', '3_pick',
       '3_win', '4_pick', '4_win', '5_pick', '5_win', '6_pick', '6_win',
       '7_pick', '7_win', '8_pick', '8_win', 'winrate', 'lane', 'main_role'],
      dtype='object')

In [4]:
df.columns
df = df[['localized_name','primary_attr','attack_type','base_health','base_mana', 'base_mana_regen', 'base_armor','base_attack_min', 'base_attack_max','base_str', 'base_agi','base_int', 'str_gain', 'agi_gain', 'int_gain', 'attack_range','move_speed','1_pick', '1_win', '2_pick', '2_win','3_pick','3_win', '4_pick', '4_win', '5_pick', '5_win', '6_pick', '6_win','7_pick', '7_win', '8_pick', '8_win', 'winrate','main_role']]
df.sample(3)

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



localized_name primary_attr attack_type  base_health  base_mana  \
115      Grimstroke          int      Ranged          200         75   
11   Phantom Lancer          agi       Melee          200         75   
54           Clinkz          agi      Ranged          200         75   

     base_mana_regen  base_armor  base_attack_min  base_attack_max  base_str  \
115              0.0         0.0               21               25        21   
11               0.0         0.0               26               40        19   
54               0.0         0.0               15               21        14   

     ...  5_pick  5_win  6_pick  6_win  7_pick  7_win  8_pick  8_win  \
115  ...   27629  13142   13809   6571    6162   2900    2168    981   
11   ...   19625   9828   10065   5050    5627   2849    2269   1152   
54   ...   15296   8393    7108   3867    3040   1734     517    279   

       winrate     main_role  
115  46.590636  hard_support  
11   50.671226      safelane  
54   56.592634           mid  

[3 rows x 35 columns]

In [5]:
df['winrate'] = round(df['winrate'], 2)

<h3>Checando valores nulos ou NaN</h3>


In [6]:
sns.heatmap(df.isna())

<AxesSubplot: >

<Figure size 1650x1050 with 2 Axes>

Não existem valores nulos ou NaN no data frame


In [7]:
df.describe()

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



base_health  base_mana  base_mana_regen  base_armor  base_attack_min  \
count        120.0      120.0       120.000000  120.000000       120.000000   
mean         200.0       75.0         0.071250    0.337500        26.850000   
std            0.0        0.0         0.208804    1.565331         7.539977   
min          200.0       75.0         0.000000   -3.000000         9.000000   
25%          200.0       75.0         0.000000   -1.000000        23.000000   
50%          200.0       75.0         0.000000    0.000000        27.000000   
75%          200.0       75.0         0.000000    1.000000        30.000000   
max          200.0       75.0         1.000000    7.000000        62.000000   

       base_attack_max    base_str    base_agi    base_int    str_gain  ...  \
count       120.000000  120.000000  120.000000  120.000000  120.000000  ...   
mean         33.791667   21.316667   18.158333   19.800000    2.705833  ...   
std           8.166231    2.992942    4.607257    3.961156    0.597234  ...   
min          11.000000   14.000000    0.000000   12.000000    1.300000  ...   
25%          29.000000   19.000000   15.000000   16.000000    2.300000  ...   
50%          34.000000   21.000000   18.000000   20.000000    2.600000  ...   
75%          38.000000   23.000000   22.000000   23.000000    3.025000  ...   
max          70.000000   29.000000   34.000000   30.000000    4.600000  ...   

              4_win        5_pick         5_win        6_pick         6_win  \
count    120.000000    120.000000    120.000000    120.000000    120.000000   
mean   16584.616667  21915.308333  10957.666667  10150.491667   5075.250000   
std    10343.131893  13227.289342   6757.259479   6124.054342   3119.665322   
min      948.000000   1378.000000    617.000000    772.000000    366.000000   
25%     9342.250000  12035.750000   6009.000000   5533.250000   2691.500000   
50%    14101.000000  19051.500000   9084.000000   8761.000000   4291.500000   
75%    21622.500000  29199.750000  14328.000000  13857.500000   6747.250000   
max    41883.000000  52388.000000  27463.000000  23344.000000  12538.000000   

             7_pick        7_win       8_pick        8_win     winrate  
count    120.000000   120.000000   120.000000   120.000000  120.000000  
mean    4429.575000  2214.791667  1326.166667   663.083333   49.902333  
std     2685.735145  1352.476085   843.289869   428.309960    2.533215  
min      421.000000   223.000000   106.000000    52.000000   43.880000  
25%     2133.750000  1044.250000   541.750000   279.000000   48.142500  
50%     3787.000000  1872.500000  1233.500000   581.500000   49.865000  
75%     6201.750000  3092.000000  1944.000000   982.000000   51.322500  
max    10048.000000  5207.000000  3307.000000  1719.000000   57.520000  

[8 rows x 31 columns]

# Visualização gráfica


In [8]:
area = plt.figure(figsize=(12,8))
fig1 = area.add_subplot(2,2,1)
fig2 = area.add_subplot(2,2,2)
fig3 = area.add_subplot(2,2,(3,4))

fig1.bar(data=df, x=df['attack_type'].unique(), height=df['attack_type'].value_counts())
fig1.set_title('Quantidade de herois por tipo de ataque',fontweight="bold")
fig1.set_ylabel('Quantidade de herois',fontweight="bold")
fig1.set_xlabel('Tipo do ataque',fontweight="bold")

fig2.bar(data=df, x=df['primary_attr'].unique(), height=df['primary_attr'].value_counts())
fig2.set_title('Quantidade de herois por atributo',fontweight="bold")
fig2.set_ylabel('Quantidade de herois',fontweight="bold")
fig2.set_xlabel('Atributo primário',fontweight="bold")

fig3.bar(data=df, x=df['main_role'].unique(), height=df['main_role'].value_counts())
fig3.set_title('Quantidade de herois por função principal',fontweight="bold")
fig3.set_ylabel('Quantidade de herois',fontweight="bold")
fig3.set_xlabel('Função principal',fontweight="bold")
area.tight_layout(pad=2.0)


<Figure size 3600x2400 with 3 Axes>

<ul>
    <li>Nota-se uma predominância de herois Melees no jogo</li>
    <li>Nota-se uma predominância de herois possuindo agilidade como seu atributo principal no jogo</li>
    <li>Safelane sendo a função principal mais realizada pelos herois e o Mid sendo a que possuem menos herois que façam bem essa função</li>
</ul>


In [9]:
df.columns

Index(['localized_name', 'primary_attr', 'attack_type', 'base_health',
       'base_mana', 'base_mana_regen', 'base_armor', 'base_attack_min',
       'base_attack_max', 'base_str', 'base_agi', 'base_int', 'str_gain',
       'agi_gain', 'int_gain', 'attack_range', 'move_speed', '1_pick', '1_win',
       '2_pick', '2_win', '3_pick', '3_win', '4_pick', '4_win', '5_pick',
       '5_win', '6_pick', '6_win', '7_pick', '7_win', '8_pick', '8_win',
       'winrate', 'main_role'],
      dtype='object')

In [10]:
df_stats = df[['primary_attr','base_health','base_mana', 'base_mana_regen', 'base_armor', 'base_attack_min',
       'base_attack_max', 'base_str', 'base_agi', 'base_int', 'str_gain',
       'agi_gain', 'int_gain', 'attack_range', 'move_speed','attack_type']]
corr = df_stats.corr().iloc[2:,2:]
corr

C:\Users\Asus\AppData\Local\Temp\ipykernel_16008\1741920496.py:4: FutureWarning:

The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



base_mana_regen  base_armor  base_attack_min  \
base_mana_regen         1.000000    0.035078        -0.023312   
base_armor              0.035078    1.000000         0.160252   
base_attack_min        -0.023312    0.160252         1.000000   
base_attack_max        -0.048143    0.129465         0.901469   
base_str               -0.176925   -0.015829         0.239328   
base_agi               -0.119268   -0.075054        -0.254760   
base_int                0.211429   -0.009351        -0.283216   
str_gain               -0.120275    0.017652         0.247829   
agi_gain               -0.060690    0.048879        -0.151789   
int_gain                0.212071   -0.104235        -0.285169   
attack_range            0.178485   -0.288959        -0.471705   
move_speed             -0.093825    0.288396        -0.037512   

                 base_attack_max  base_str  base_agi  base_int  str_gain  \
base_mana_regen        -0.048143 -0.176925 -0.119268  0.211429 -0.120275   
base_armor              0.129465 -0.015829 -0.075054 -0.009351  0.017652   
base_attack_min         0.901469  0.239328 -0.254760 -0.283216  0.247829   
base_attack_max         1.000000  0.209702 -0.288803 -0.198214  0.266800   
base_str                0.209702  1.000000 -0.317515 -0.115112  0.540538   
base_agi               -0.288803 -0.317515  1.000000 -0.063175 -0.404380   
base_int               -0.198214 -0.115112 -0.063175  1.000000 -0.194869   
str_gain                0.266800  0.540538 -0.404380 -0.194869  1.000000   
agi_gain               -0.216543 -0.350574  0.659817 -0.192250 -0.359609   
int_gain               -0.208316 -0.288821 -0.098293  0.623221 -0.380693   
attack_range           -0.394171 -0.479830  0.004906  0.436949 -0.474264   
move_speed             -0.054764 -0.003106  0.145364  0.019970  0.017994   

                 agi_gain  int_gain  attack_range  move_speed  
base_mana_regen -0.060690  0.212071      0.178485   -0.093825  
base_armor       0.048879 -0.104235     -0.288959    0.288396  
base_attack_min -0.151789 -0.285169     -0.471705   -0.037512  
base_attack_max -0.216543 -0.208316     -0.394171   -0.054764  
base_str        -0.350574 -0.288821     -0.479830   -0.003106  
base_agi         0.659817 -0.098293      0.004906    0.145364  
base_int        -0.192250  0.623221      0.436949    0.019970  
str_gain        -0.359609 -0.380693     -0.474264    0.017994  
agi_gain         1.000000 -0.274888     -0.002032   -0.030226  
int_gain        -0.274888  1.000000      0.671435   -0.157118  
attack_range    -0.002032  0.671435      1.000000   -0.245683  
move_speed      -0.030226 -0.157118     -0.245683    1.000000

In [11]:
matrix = np.triu(corr)
sns.heatmap(corr,mask=matrix)
plt.show()

<Figure size 1650x1050 with 2 Axes>

Existem algumas informações importantes nesse gráfico.
<ol>
<li>Percebe-se que quanto maior o atributo básico(provavelmente o atributo primário do heroi) maior o seu ganho ao longo do jogo</li>
<li>o ataque básico mínimo e o máximo tem uma grande correlação (0.901469), indicando que os 2 tendem a variar juntos, quanto maior o ataque máximo, maior o ataque mínimo, tendendo a seguir o mesmo delta(intervalo)</li>
<li>A mais curiosa é sem duvida a correlação relevante entre o atributo de inteligência e o 'attack range', uma das razões disso se da pela maioria dos herois de inteligência serem herois que atacam de 'ranged'</li>
<li>Contrariamente aos herois de inteligência, os herois de força em sua grande maioria são 'melees', fazendo assim com que esses possuam um 'attack range' menor</li>
</ol>
<h4>gráfico abaixo servem para fundamentar os tópicos 3 e 4</h4>


In [12]:
df['primary_attr'].unique()

array(['agi', 'str', 'int'], dtype=object)

In [13]:
df_str_int = df[(df['primary_attr'] == 'str') | (df['primary_attr'] == 'int')]
sns.countplot(x = 'primary_attr', hue = 'attack_type', data = df_str_int)
plt.title('Quantidade de herois por tipo de ataque', fontweight="bold")
plt.xlabel('Atributo primário')
plt.ylabel('Quantidade de herois')
plt.show()

<Figure size 1650x1050 with 1 Axes>

# Insights textuais


In [14]:
first_pick = df[df['1_pick'] == df['1_pick'].max()]['localized_name'].values
second_pick = df[df['2_pick'] == df['2_pick'].max()]['localized_name'].values
third_pick = df[df['3_pick'] == df['3_pick'].max()]['localized_name'].values
fourth_pick = df[df['4_pick'] == df['4_pick'].max()]['localized_name'].values
fifth_pick = df[df['5_pick'] == df['5_pick'].max()]['localized_name'].values
sixth_pick = df[df['6_pick'] == df['6_pick'].max()]['localized_name'].values
seventh_pick = df[df['7_pick'] == df['7_pick'].max()]['localized_name'].values
eighth_pick = df[df['8_pick'] == df['8_pick'].max()]['localized_name'].values
list = [first_pick,second_pick,third_pick,fourth_pick,fifth_pick,sixth_pick,seventh_pick,eighth_pick]
for i in range(0,8):
    print(f'O heroi mais escolhido na {i+1} escolha é {list[i][0]}')

O heroi mais escolhido na 1 escolha é Sniper
O heroi mais escolhido na 2 escolha é Pudge
O heroi mais escolhido na 3 escolha é Pudge
O heroi mais escolhido na 4 escolha é Pudge
O heroi mais escolhido na 5 escolha é Pudge
O heroi mais escolhido na 6 escolha é Juggernaut
O heroi mais escolhido na 7 escolha é Juggernaut
O heroi mais escolhido na 8 escolha é Void Spirit


criação de nova coluna para saber qual é o heroi mais escolhido, independente de posição ou ordem da escolha.


In [15]:
df['quantidade_vezes_escolhido'] = df['1_pick'] + df['2_pick'] + df['3_pick'] + df['4_pick'] + df['5_pick'] + df['6_pick'] + df['7_pick'] + df['8_pick']
herois_mais_escolhidos = df.sort_values(by='quantidade_vezes_escolhido', ascending=False).head(3)[['localized_name', 'quantidade_vezes_escolhido']]
herois_mais_escolhidos.index = range(1, herois_mais_escolhidos.shape[0] + 1)
herois_mais_escolhidos

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



localized_name  quantidade_vezes_escolhido
1          Pudge                      334420
2     Juggernaut                      322241
3           Lion                      318075

2 Herois (Pudge e Juggernaut) apareceram na lista anterior
Importante também saber os menos escolhidos


In [16]:
herois_menos_escolhidos = df.sort_values(by='quantidade_vezes_escolhido', ascending=True).head(3)[['localized_name', 'quantidade_vezes_escolhido']]
herois_menos_escolhidos.index = range(1, herois_menos_escolhidos.shape[0] + 1)
herois_menos_escolhidos

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



localized_name  quantidade_vezes_escolhido
1           Chen                        9252
2         Visage                       16768
3     Brewmaster                       24069

Estudando agora os herois com maior e menor winrate


In [17]:
media_winrate = round(df['winrate'].mean(),2)
print(f'Sabendo que a média de winrate é {media_winrate}')

Sabendo que a média de winrate é 49.9


In [18]:
herois_maior_winrate = df.sort_values(by='winrate', ascending=False).head(3)[['localized_name', 'winrate']]
herois_maior_winrate.index = range(1, herois_maior_winrate.shape[0] + 1)
round(herois_maior_winrate,2)

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



localized_name  winrate
1          Lycan    57.52
2    Broodmother    56.86
3         Clinkz    56.59

MENOR


In [19]:
herois_maior_winrate = df.sort_values(by='winrate', ascending=True).head(3)[['localized_name', 'winrate']]
herois_maior_winrate.index = range(1, herois_maior_winrate.shape[0] + 1)
round(herois_maior_winrate,2)

C:\Users\Asus\Desktop\Heroes Dota2\venvdota\lib\site-packages\IPython\core\formatters.py:342: FutureWarning:

In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.



localized_name  winrate
1           Sven    43.88
2         Magnus    44.47
3           Tiny    45.48

# Visualização gráfica 2


In [20]:
fig = px.scatter(data_frame=df, 
                 x='main_role', 
                 y='winrate', 
                 color='primary_attr',
                 marginal_y="box",
                 hover_data=['main_role','winrate','primary_attr','localized_name'],
                 title="<b> Taxa de vitória por posição principal dos herois </b>")
fig.show()

<h6>Usar esse gráfico para fazer dashboard</h6>
<ul>
<li>Safelane tem em sua maioria herois de agilidade com essa posição principal</li>
<li>Hard_support tem em sua maioria herois de inteligência com essa posição principal</li>
<li>Offlane tem em sua maioria herois de força com essa posição principal</li>
</ul>


In [21]:
df.to_csv('./data/heroes_data_limpo.csv')